frontend/catalyst/api_extensions/quantum_passes.py

```diff
@@ -0,0 +1,156 @@
+                qml.Hadamard(wires=0)
+                qml.Hadamard(wires=0)
+                return qml.expval(qml.PauliZ(0))
+
+            @qml.qnode(qml.device("lightning.qubit", wires=1))
+            def g(x: float):
+                qml.RX(x, wires=0)
+                qml.Hadamard(wires=0)
+                qml.Hadamard(wires=0)
+                return qml.expval(qml.PauliZ(0))
+
+            ff = f(1.0)
+            gg = g(2.0)
+
+            return ff, gg
+
+    >>> workflow()
+    (Array(0.54030231, dtype=float64), Array(-0.41614684, dtype=float64))
+
+    In the compiled mlir files, specifically in ``workflow/2_QuantumCompilationPass.mlir``:
+
+    .. code-block:: mlir
+
+          func.func private @f(%arg0: tensor<f64>) -> tensor<f64> {
+            quantum.device["PATH_TO_librtd_lightning.so", "LightningSimulator",
+            "{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}"]
+            %0 = quantum.alloc( 1) : !quantum.reg
+            %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
+            %extracted = tensor.extract %arg0[] : tensor<f64>
+            %out_qubits = quantum.custom "RX"(%extracted) %1 : !quantum.bit
+            %2 = quantum.namedobs %out_qubits[ PauliZ] : !quantum.obs
+            %3 = quantum.expval %2 : f64
+            %from_elements = tensor.from_elements %3 : tensor<f64>
+            %4 = quantum.insert %0[ 0], %out_qubits : !quantum.reg, !quantum.bit
+            quantum.dealloc %4 : !quantum.reg
+            quantum.device_release
+            return %from_elements : tensor<f64>
+          }
+          func.func private @g(%arg0: tensor<f64>) -> tensor<f64> {
+            quantum.device["PATH_TO_librtd_lightning.so", "LightningSimulator",
+            "{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}"]
+            %0 = quantum.alloc( 1) : !quantum.reg
+            %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
+            %extracted = tensor.extract %arg0[] : tensor<f64>
+            %out_qubits = quantum.custom "RX"(%extracted) %1 : !quantum.bit
+            %out_qubits_0 = quantum.custom "Hadamard"() %out_qubits : !quantum.bit
+            %out_qubits_1 = quantum.custom "Hadamard"() %out_qubits_0 : !quantum.bit
+            %2 = quantum.namedobs %out_qubits_1[ PauliZ] : !quantum.obs
+            %3 = quantum.expval %2 : f64
+            %from_elements = tensor.from_elements %3 : tensor<f64>
+            %4 = quantum.insert %0[ 0], %out_qubits_1 : !quantum.reg, !quantum.bit
+            quantum.dealloc %4 : !quantum.reg
+            quantum.device_release
+            return %from_elements : tensor<f64>
+          }
+
+    We see that ``f``, decorated by ``cancel_inverses``, no longer has the neighbouring Hadamard gates.
+    """
+    if not isinstance(fn, qml.QNode):
+        raise TypeError(f"A QNode is expected, got the classical function {fn}")
+
+    wrapped_qnode_function = fn.func
+
+    def wrapper(*args, **kwrags):
```

Comment:
One issue with this wrapping strategy is that repeated calls to the same qnode in a qjit will generate multiple apply_pass primitives that are the same.

Reply:
Great point :/ In that case I'm not sure what would be the best way to allow defining qnodes at the module level

Reply:
(Jotting down a thought on Friday night)

In Python mutable default argument values are shared across all calls. Using this fact there's this cute little trick to keep track of whether a function has been called before 

```Python
def f(first_call = [True]):
    print("entering, ", first_call)
    if first_call[0]:
        print("first call!\n")
    first_call[0] =False
    print("leaving, ", first_call)


f()
f()
f()

>>>
entering,  [True]
first call!

leaving,  [False]
entering,  [False]
leaving,  [False]
entering,  [False]
leaving,  [False]
```

So we can make only the first call of the wrapper do the apply_pass.bind(). But since Boolean values of arguments are used for comparisons, not sure if this is traceable. 

On the other hand this fist_call default argument isn't actually a dynamical user argument and is only here to help tracing to jaxpr, so there might be ways to make it work.